edb/lib/cfg.edgeql

```diff
@@ -102,6 +102,71 @@ CREATE TYPE cfg::Auth EXTENDING cfg::ConfigObject {
     };
 };
 
+CREATE SCALAR TYPE cfg::SMTPSecurity EXTENDING enum<
+    PlainText,
+    TLS,
+    STARTTLS,
+    STARTTLSOrPlainText,
+>;
+
+CREATE ABSTRACT TYPE cfg::EmailProviderConfig EXTENDING cfg::ConfigObject {
+    CREATE REQUIRED PROPERTY name -> std::str {
+        CREATE CONSTRAINT std::exclusive;
+        CREATE ANNOTATION std::description :=
+            "The name of the email provider.";
+    };
+};
+
+CREATE TYPE cfg::SMTPProviderConfig EXTENDING cfg::EmailProviderConfig {
+    CREATE PROPERTY sender -> std::str {
+        CREATE ANNOTATION std::description :=
+            "\"From\" address of system emails sent for e.g. \
+            password reset, etc.";
+    };
+    CREATE PROPERTY host -> std::str {
+        CREATE ANNOTATION std::description :=
+            "Host of SMTP server to use for sending emails. \
+            If not set, \"localhost\" will be used.";
+    };
+    CREATE PROPERTY port -> std::int32 {
+        CREATE ANNOTATION std::description :=
+            "Port of SMTP server to use for sending emails. \
+            If not set, common defaults will be used depending on security: \
+            465 for TLS, 587 for STARTTLS, 25 otherwise.";
+    };
+    CREATE PROPERTY username -> std::str {
+        CREATE ANNOTATION std::description :=
+            "Username to login as after connected to SMTP server.";
+    };
+    CREATE PROPERTY password -> std::str {
+        SET secret := true;
+        CREATE ANNOTATION std::description :=
+            "Password for login after connected to SMTP server.";
+    };
+    CREATE REQUIRED PROPERTY security -> cfg::SMTPSecurity {
+        SET default := cfg::SMTPSecurity.STARTTLSOrPlainText;
+        CREATE ANNOTATION std::description :=
+            "Security mode of the connection to SMTP server. \
+            By default, initiate a STARTTLS upgrade if supported by the \
+            server, or fallback to PlainText.";
+    };
+    CREATE REQUIRED PROPERTY validate_certs -> std::bool {
+        SET default := true;
+        CREATE ANNOTATION std::description :=
+            "Determines if SMTP server certificates are validated.";
+    };
+    CREATE REQUIRED PROPERTY timeout_per_email -> std::duration {
+        SET default := <std::duration>'60 seconds';
+        CREATE ANNOTATION std::description :=
+            "Maximum time to send an email, including retry attempts.";
+    };
+    CREATE REQUIRED PROPERTY timeout_per_attempt -> std::duration {
+        SET default := <std::duration>'15 seconds';
+        CREATE ANNOTATION std::description :=
+            "Maximum time for each SMTP request.";
+    };
```

Comment:
Max concurrency config here?

Reply:
Yeah, I think so. Though this would become a per-provider config, instead of a server/tenant-wide config. It should be fine for now, but in the long run we may want to have both: per-provider config makes sure of not overwhelming the provider, while the server/tenant-wide config makes sure e.g. the local number of FDs is not exceeded.